smartmeter_datacollector/smartmeter/siemens_td3511.py

```diff
@@ -0,0 +1,255 @@
+    RegisterDataPoint("2.8.2", MeterDataPointTypes.ACTIVE_ENERGY_N_T2.value, 1000),
+    RegisterDataPoint("3.8.1", MeterDataPointTypes.REACTIVE_ENERGY_P.value, 1000),
+    RegisterDataPoint("4.8.1", MeterDataPointTypes.REACTIVE_ENERGY_N.value, 1000),
+]
+
+
+class SiemensParser():
+    REGEX = r"(.{3,20})\(([\d\-\.:]{3,20})[*\)](.{0,10}[^\)\r\n])?"
+
+    def __init__(self, use_system_time: bool = False) -> None:
+        self._use_system_time = use_system_time
+        self._timestamp = None
+        self._meter_id = None
+        self._buffer = []
+        self._register_obis = {r.obis: r for r in DEFAULT_REGISTER_MAPPING}
+
+    def append_to_buffer(self, received_data):
+        self._buffer.append(received_data.decode())
+
+    def clear_buffer(self):
+        self._buffer = []
+
+    def parse_data_objects(self, timestamp: datetime):
+        # Extract timestamp and meter id
+        self._timestamp = timestamp
+        meter_time = None
+        meter_date = None
+        for data in self._buffer:
+            result = re.search(SiemensParser.REGEX, data)
+            if result is None:
+                continue
+            obis, value, _ = result.groups()
+
+            # Extract meter id (common source id for all data points)
+            if obis == "0.0.0":
+                self._meter_id = value
+            # Extract date and time
+            try:
+                if obis == "0.9.1":
+                    meter_time = datetime.strptime(value, "%H:%M:%S").time()
+                if obis == "0.9.2":
+                    meter_date = datetime.strptime(value, "%y-%m-%d").date()
+            except ValueError:
+                meter_time = None
+                meter_date = None
+                LOGGER.warning("Invalid timestamp received: %s. Using system time instead.", value)
+            if meter_date is not None and meter_time is not None and not self._use_system_time:
+                self._timestamp = datetime.combine(meter_date, meter_time).astimezone(timezone.utc)
+
+        # Extract register data
+        data_points: List[MeterDataPoint] = []
+        for data in self._buffer:
+            result = re.search(SiemensParser.REGEX, data)
+            if result is None:
+                continue
+            obis, value, _ = result.groups()
+
+            if value is None:
+                LOGGER.warning("No value received for %s.", obis)
+                continue
+
+            reg_type = self._register_obis.get(obis, None)
+            if reg_type is None:
+                continue
+            data_point_type = reg_type.data_point_type
+
+            try:
+                scaled_value = float(value) * reg_type.scaling
+            except (TypeError, ValueError, OverflowError):
+                LOGGER.warning("Invalid register value '%s'. Skipping register.", str(value))
+                continue
+
+            data_points.append(MeterDataPoint(data_point_type, scaled_value, self._meter_id, self._timestamp))
```

Comment:
`self._meter_id` can be `None`. Handle this case by using some kind of a fallback ID like serial port or fixed string like meter model type.

Reply:
I think setting it to 'unknown' makes most sense if the meter id is really missing.